src/email-provider/mail-provider.service.ts

```diff
@@ -0,0 +1,67 @@
+import { Injectable } from '@nestjs/common';
+import { MailProvider } from './types/email-provider.interface';
+import { MailgunProvider } from './providers/mailgun-provider';
+import { SmtpProvider } from './providers/smtp-provider';
+import { SendGridProvider } from './providers/sendgrid-provider';
+import {
+    MailgunConfig,
+    SendgridConfig,
+    SmtpConfig,
+} from './types/provider-config.interface';
+import { StorageService } from 'src/storage/storage.service';
+
+@Injectable()
+export class EmailProviderService {
+    private providers = new Map<string, MailProvider>();
+
+    constructor(private readonly storage: StorageService) {
+        this.providers.set('mailgun', new MailgunProvider(storage)),
+            this.providers.set('smtp', new SmtpProvider(storage));
+        this.providers.set('sendgrid', new SendGridProvider(storage));
+    }
+
+    async sendMail(
+        from: string,
+        to: string,
+        subject: string,
+        template: string,
+        providerName: string,
+        providerConfig: MailgunConfig | SendgridConfig | SmtpConfig,
+    ) {
+        const provider = this.providers.get(providerName);
+
+        if (provider) {
```

Comment:
tutaj zrobilbym odwrotna logike - jezeli nie ma providera to po prostu wyrzucil blad a nie wykonywal akcje jezeli ten provider jest